### ISSUE TITLE
Problem with iThreads, and perl 5.8.1 & 5.8.2 RC1, with hashes

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6888]
Migrated from [rt.perl.org#24383](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=24383) (status was 'resolved')

Searchable as RT24383$

**Comment by p5pRT** [comment]
### From will@rrtrain.ath.cx

#### Created by will@rrtrain.ath.cx



Using Perl versions 5\.8\.1 or 5\.8\.2\, compiled with iThreads

I have a custom module that returns a hash\.  It is called from
within a thread\.  When done under 5\.8\.1 or 5\.8\.2 it returns the 
following errors \(which did not occur with 5\.8\.0\)&#8203;:

Unbalanced string table refcount&#8203;: \(3\) for "IO&#8203;::Select" during global destruction\.
Unbalanced string table refcount&#8203;: \(3\) for "Socket" during global destruction\.
Attempt to free non\-existent shared string 'IO&#8203;::Select' during global destruction\.
Attempt to free non\-existent shared string 'Socket' during global destruction\.
Attempt to free non\-existent shared string 'IO&#8203;::Select' during global destruction\.
Attempt to free non\-existent shared string 'Socket' during global destruction\.
Attempt to free non\-existent shared string 'IO&#8203;::Select' during global destruction\.
Attempt to free non\-existent shared string 'Socket' during global destruction\.

This does not occur if the module is not called from a thread\.

I followed the thread at
http&#8203;://lists\.debian\.org/debian\-perl/2003/debian\-perl\-200310/msg00020\.html
which lead me to attempt to set PERL\_HASH\_SEED=0\, but the issue still exists\.

This issue is of high severity for me as it causes mod\_perl to segfault\.  

Here is my perl configuration&#8203;:

Summary of my perl5 \(revision 5\.0 version 8 subversion 2\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.4\.21\-0\.13mdk\, archname=i686\-linux\-thread\-multi
&nbsp;    uname='linux rrtrain\.ath\.cx 2\.4\.21\-0\.13mdk \#1 fri mar 14 15&#8203;:08&#8203;:06 est 2003 i686 unknown unknown gnulinux '
&nbsp;    config\_args=''
&nbsp;    hint=previous\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=undef usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-DNO\_HASH\_SEED'\,
&nbsp;    optimize='\-O3'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DTHREADS\_HAVE\_PIDS \-fno\-strict\-aliasing \-I/usr/local/include \-DNO\_HASH\_SEED'
&nbsp;    ccversion=''\, gccversion='3\.2\.2 \(Mandrake Linux 9\.1 3\.2\.2\-3mdk\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=4
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lgdbm \-ldl \-lm \-lcrypt \-lutil \-lpthread \-lc
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lpthread \-lc
&nbsp;    libc=/lib/libc\-2\.3\.1\.so\, so=so\, useshrplib=true\, libperl=libperl\.so
&nbsp;    gnulibc\_version='2\.3\.1'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic \-Wl\,\-rpath\,/usr/lib/perl5/5\.8\.2/i686\-linux\-thread\-multi/CORE'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: MULTIPLICITY USE\_ITHREADS PERL\_IMPLICIT\_CONTEXT
&nbsp;  Locally applied patches&#8203;:
&nbsp;        RC1
&nbsp;  Built under linux
&nbsp;  Compiled at Oct 31 2003 21&#8203;:48&#8203;:24
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /usr/lib/perl5/5\.8\.2/i686\-linux\-thread\-multi
&nbsp;    /usr/lib/perl5/5\.8\.2
&nbsp;    /usr/lib/perl5/site\_perl/5\.8\.2/i686\-linux\-thread\-multi
&nbsp;    /usr/lib/perl5/site\_perl/5\.8\.2
&nbsp;    /usr/lib/perl5/site\_perl
&nbsp;    \.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.2:

Configured by will at Fri Oct 31 20:49:18 CST 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 2) configuration:
  Platform:
    osname=linux, osvers=2.4.21-0.13mdk, archname=i686-linux-thread-multi
    uname='linux rrtrain.ath.cx 2.4.21-0.13mdk #1 fri mar 14 15:08:06 est 2003 i686 unknown unknown gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=undef usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -fno-strict-aliasing -I/usr/local/include',
    optimize='-O3',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='3.2.2 (Mandrake Linux 9.1 3.2.2-3mdk)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc-2.3.1.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.3.1'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/usr/lib/perl5/5.8.2/i686-linux-thread-multi/CORE'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    RC1


@INC for perl v5.8.2:
    /usr/lib/perl5/5.8.2/i686-linux-thread-multi
    /usr/lib/perl5/5.8.2
    /usr/lib/perl5/site_perl/5.8.2/i686-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.2
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.8.2:
    HOME=/root
    LANG=en_US
    LANGUAGE=en_US:en
    LC_ADDRESS=en_US
    LC_COLLATE=en_US
    LC_CTYPE=en_US
    LC_IDENTIFICATION=en_US
    LC_MEASUREMENT=en_US
    LC_MESSAGES=en_US
    LC_MONETARY=en_US
    LC_NAME=en_US
    LC_NUMERIC=en_US
    LC_PAPER=en_US
    LC_TELEPHONE=en_US
    LC_TIME=en_US
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/sbin:/usr/sbin:/bin:/usr/bin:/usr/X11R6/bin:/usr/local/bin:/usr/local/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From sky@nanisky.com


On Saturday\, November 1\, 2003\, at 03&#8203;:53  am\, will@&#8203;rrtrain\.ath\.cx \(via 
RT\) wrote&#8203;:

> Using Perl versions 5\.8\.1 or 5\.8\.2\, compiled with iThreads
>
> I have a custom module that returns a hash\.  It is called from
> within a thread\.  When done under 5\.8\.1 or 5\.8\.2 it returns the
> following errors \(which did not occur with 5\.8\.0\)&#8203;:
>
> Unbalanced string table refcount&#8203;: \(3\) for "IO&#8203;::Select" during global 
> destruction\.
> Unbalanced string table refcount&#8203;: \(3\) for "Socket" during global 
> destruction\.
> Attempt to free non\-existent shared string 'IO&#8203;::Select' during global 
> destruction\.
> Attempt to free non\-existent shared string 'Socket' during global 
> destruction\.
> Attempt to free non\-existent shared string 'IO&#8203;::Select' during global 
> destruction\.
> Attempt to free non\-existent shared string 'Socket' during global 
> destruction\.
> Attempt to free non\-existent shared string 'IO&#8203;::Select' during global 
> destruction\.
> Attempt to free non\-existent shared string 'Socket' during global 
> destruction\.
>
> This does not occur if the module is not called from a thread\.
>

Hello\,

To be able to debug this\, we would need a test script \(as small as 
possible\) that reproduces the bug\, otherwise we are kind of lost in the 
dark\.

With 'module is not called' do you mean a subroutine in the module is 
called or do you mean the module was required?

Arthur

**Comment by p5pRT** [comment]
### From will@rrtrain.ath.cx

Hey\, sorry for the semi\-incoherent message\, I'd never used perlbug before\, and I didn't realize it was going to email live people\.  Below is the test script that I use to reproduce the error\.  

Here's what I've discovered\, and it has nothing to do with hashes as my original message indicated\.  Instead it appears to be something between 5\.8\.2's iThreads\, and LWP&#8203;::UserAgent \(Version 2\.23 built on 10/26/03\)\.

Whenever push a thread that calls LWP&#8203;::UserAgent\, just to retrieve a web page for example\, and then close that thread\, I recieve&#8203;:

Unbalanced string table refcount&#8203;: \(3\) for "Socket" during global destruction\.
Unbalanced string table refcount&#8203;: \(3\) for "IO&#8203;::Select" during global destruction\.
Attempt to free non\-existent shared string 'IO&#8203;::Select' during global destruction\.
Attempt to free non\-existent shared string 'Socket' during global destruction\.
Attempt to free non\-existent shared string 'IO&#8203;::Select' during global destruction\.
Attempt to free non\-existent shared string 'Socket' during global destruction\.
Attempt to free non\-existent shared string 'IO&#8203;::Select' during global destruction\.
Attempt to free non\-existent shared string 'Socket' during global destruction\.


If I call the sub routine outside of the thread\, it works fine\.  Only when it is run from within a thread does it break\.

This occured with 5\.8\.1 and 5\.8\.2\.

Thanks for all your help\!

\#\!/usr/bin/perl
use strict;
use LWP&#8203;::UserAgent;
use threads;
use threads&#8203;::shared;
my $request="http&#8203;://www\.google\.com/index\.html";
my $exampleThread=threads\->new \(\\&lwpAgent\,$request\);
my $content=$exampleThread\->join\(\);
print "here is the Error&#8203;:\\n\\n";
exit 0;

sub lwpAgent\{
&nbsp;        my \($request\)=@&#8203;\_;
&nbsp;        my $ua = LWP&#8203;::UserAgent\->new;
&nbsp;        $ua\->agent\("ModemChecker\.cgi"\);
&nbsp;        my $req = HTTP&#8203;::Request\->new\(GET => $request\);
&nbsp;        $req\->content\('match=www&errors=0'\);
&nbsp;        my $res = $ua\->request\($req\);
&nbsp;        my $content = $res\->content;
&nbsp;        return $content;
\}

Will Berger

**Comment by p5pRT** [comment]
### From tobeya@tobert.org

Subject is from this thread I found on google&#8203;:
http&#8203;://groups\.google\.com/groups?hl=en&lr=&ie=UTF\-8&th=5f61fd7e8929bab9&rnum=4

Test script is attached or see http&#8203;://www\.tobert\.org/tmp/test\.txt\. 
Comment out the "Timeout" option for IO&#8203;::Socket and the problem goes
away\.  perl 5\.8\.2 \(custom build\) gives the following output&#8203;:

tobeya@&#8203;localhost> perl test\.pl
\.\.\.
Unbalanced string table refcount&#8203;: \(3\) for "IO&#8203;::Select" during global destruction\.
Attempt to free non\-existent shared string 'IO&#8203;::Select' during global destruction\.
Attempt to free non\-existent shared string 'IO&#8203;::Select' during global destruction\.
Attempt to free non\-existent shared string 'IO&#8203;::Select' during global destruction\.

I'll test patches if there are any\.  I'm not subscribe to p5p\, so please
CC me\.

OS is Red Hat Linux Advanced Workstation 3\.0 running a 2\.6\.0\-test6
kernel and glibc\-2\.3\.2\-95\.6\.  Red Hat's perl 5\.8\.1 on the same box also
crashes with the same error\.  Also shows up on RH Enterprise Server with
a RH kernel and my perl 5\.8\.2 build\, but not with the RH 5\.8\.0 rpm\. 
Does not show up on my Red Hat 7\.3 box with perl 5\.8\.0\.

I suspect it's something introduced in 5\.8\.\[12\] and may be related to
differences between LinuxThreads and NPTL\.

\-Al Tobey


---
perl \-V

Summary of my perl5 \(revision 5\.0 version 8 subversion 2\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.4\.21\-4\.0\.1\.elsmp\, archname=i686\-linux\-thread\-multi
&nbsp;    uname='linux loach 2\.4\.21\-4\.0\.1\.elsmp \#1 smp thu oct 23 01&#8203;:27&#8203;:36 edt 2003 i686 i686 i386 gnulinux '
&nbsp;    config\_args='\-der'
&nbsp;    hint=previous\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O3'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fno\-strict\-aliasing \-I/usr/local/include \-D\_REENTRANT \-D\_GNU\_SOURCE \-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64 \-D\_REENTRANT \-D\_GNU\_SOURCE \-fno\-strict\-aliasing \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'
&nbsp;    ccversion=''\, gccversion='3\.2\.3 20030502 \(Red Hat Linux 3\.2\.3\-20\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lpthread \-lc
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lpthread \-lc
&nbsp;    libc=/lib/libc\-2\.3\.2\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.3\.2'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: MULTIPLICITY USE\_ITHREADS USE\_LARGE\_FILES PERL\_IMPLICIT\_CONTEXT
&nbsp;  Built under linux
&nbsp;  Compiled at Dec  5 2003 12&#8203;:38&#8203;:28
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /opt/perl\-5\.8\.2/lib/5\.8\.2/i686\-linux\-thread\-multi
&nbsp;    /opt/perl\-5\.8\.2/lib/5\.8\.2
&nbsp;    /opt/perl\-5\.8\.2/lib/site\_perl/5\.8\.2/i686\-linux\-thread\-multi
&nbsp;    /opt/perl\-5\.8\.2/lib/site\_perl/5\.8\.2
&nbsp;    /opt/perl\-5\.8\.2/lib/site\_perl
&nbsp;    \.

**Comment by p5pRT** [comment]
### From tobeya@tobert.org

[test.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/68883/51236/test.pl)

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Dec 23\, 2003 at 05&#8203;:53&#8203;:52PM \-0500\, Al Tobey wrote&#8203;:
> Subject is from this thread I found on google&#8203;:
> http&#8203;://groups\.google\.com/groups?hl=en&lr=&ie=UTF\-8&th=5f61fd7e8929bab9&rnum=4
> 
> Test script is attached or see http&#8203;://www\.tobert\.org/tmp/test\.txt\. 
> Comment out the "Timeout" option for IO&#8203;::Socket and the problem goes
> away\.  perl 5\.8\.2 \(custom build\) gives the following output&#8203;:
> 
> tobeya@&#8203;localhost> perl test\.pl
> \.\.\.
> Unbalanced string table refcount&#8203;: \(3\) for "IO&#8203;::Select" during global destruction\.
> Attempt to free non\-existent shared string 'IO&#8203;::Select' during global destruction\.
> Attempt to free non\-existent shared string 'IO&#8203;::Select' during global destruction\.
> Attempt to free non\-existent shared string 'IO&#8203;::Select' during global destruction\.

This is an interesting bug\. It can be reduced to the following test case&#8203;:

&nbsp;    use threads;
&nbsp;    our %foo : unique;
&nbsp;    sub f \{ $foo\{baz\}=0; \}
&nbsp;    my $thr = threads\->create\( \\&f \);
&nbsp;    $thr\->join\(\);

What's happening is that the :unique attribute is supposed to cause
everything in \*foo \(including %foo\) to turn readonly when the first
interpreter is cloned; however due to a trivial cut\+paste typo in
sv\.c&#8203;:S\_gv\_share\(\)\, the HV in the typeglob never gets marked RO\. I can fix
this trivially; however it has two side\-effects\.

Firstly\, once %foo gets marked readonly in the cloned thread\, Perl thinks
that it's a restricted hash\, so the script above now produces the run\-time
error&#8203;:

&nbsp;    thread failed to start&#8203;: Attempt to access disallowed key 'baz' in a
&nbsp;    restricted hash at /tmp/p2 line 5\.

which might be regarded as confusing or misleading\.

Second\, the original code involving IO&#8203;::Socket triggered this bug because
one thread ended up require'ing IO&#8203;::Select\, which then did a
C\<use warnings&#8203;::register>\, which has the effect of modifying some :unique
hashes defined in the C\<warnings> package\. So\, fixing this bug may cause
IO&#8203;::Select \(and stuff which invokes it\) to die unexpectedly\. There may be
other similar code\.

So\, should I fix this bug and at the same time remove the :unique'
attribute from all the hashes in warnings\.pm?

Dave\.

\-\- 
"I do not resent critisism\, even when\, for the sake of emphasis\,
it parts for the time with reality"\.
&nbsp;    \-\- Winston Churchill\, House of Commons\, 22nd Jan 1941\.

**Comment by p5pRT** [comment]
### From @lizmat

At 23&#8203;:25 \+0000 12/31/03\, Dave Mitchell wrote&#8203;:
>So\, should I fix this bug and at the same time remove the :unique'
>attribute from all the hashes in warnings\.pm?

I think the latter should be done anyway&#8203;: The "&#8203;:unique" attribute is 
supposed to be used only in those cases where you can be sure that 
the entity will remain unchanged after cloning\.  This is \_clearly\_ 
not the case with warnings\.pm\, and should therefore be removed\.


Liz

**Comment by p5pRT** [comment]
### From @iabyn

On Thu\, Jan 01\, 2004 at 11&#8203;:29&#8203;:39AM \+0100\, Elizabeth Mattijsen wrote&#8203;:
> At 23&#8203;:25 \+0000 12/31/03\, Dave Mitchell wrote&#8203;:
> >So\, should I fix this bug and at the same time remove the :unique'
> >attribute from all the hashes in warnings\.pm?
> 
> I think the latter should be done anyway&#8203;: The "&#8203;:unique" attribute is 
> supposed to be used only in those cases where you can be sure that 
> the entity will remain unchanged after cloning\.  This is \_clearly\_ 
> not the case with warnings\.pm\, and should therefore be removed\.

Change 22034 by davem@&#8203;davem\-percy on 2004/01/01 19&#8203;:58&#8203;:08

&nbsp;    Fix bug \#24383\, where hashes with the :unique attribute weren't
&nbsp;    getting made readonly on interpreter clone\. Also\, remove the
&nbsp;    :unique attribute from the hashes in warnings\.pm\, since they may
&nbsp;    later be modified by warnings&#8203;::register\.  Finally\, adds tests for
&nbsp;    the :unique attribute\.

\-\- 
"Emacs isn't a bad OS once you get used to it\.
It just lacks a decent editor\."

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
